Kernel/Process.cpp

```diff
@@ -3053,18 +3053,33 @@ int Process::sys$select(const Syscall::SC_select_params* params)
     return marked_fd_count;
 }
 
-int Process::sys$poll(pollfd* fds, int nfds, int timeout)
+int Process::sys$poll(const Syscall::SC_poll_params* params)
 {
     REQUIRE_PROMISE(stdio);
-    if (!validate_read_typed(fds))
+    // FIXME: Return -EINVAL if timeout is invalid.
+    if (!validate_read_typed(params))
         return -EFAULT;
 
     SmapDisabler disabler;
 
+    pollfd* fds = params->fds;
+    unsigned long nfds = params->nfds;
+    const timespec* timeout = params->timeout;
+    const sigset_t* sigmask = params->sigmask;
+
+    if (fds && !validate_read_typed(fds, nfds))
+        return -EFAULT;
+    if (timeout && !validate_read_typed(timeout))
+        return -EFAULT;
+    if (sigmask && !validate_read_typed(sigmask))
+        return -EFAULT;
+    if (!validate_read_typed(fds))
+        return -EFAULT;
+
     Thread::SelectBlocker::FDVector rfds;
     Thread::SelectBlocker::FDVector wfds;
 
-    for (int i = 0; i < nfds; ++i) {
+    for (unsigned long i = 0; i < nfds; ++i) {
```

Comment:
~We usually~ Unixes of all kinds use `int` for fds, so why use `unsigned long` here?

Reply:
Hmm, probably because of `nfds_t`. Then why not use that here? TBH I'd prefer an `int` nevertheless

Reply:
https://pubs.opengroup.org/onlinepubs/009695399/basedefs/poll.h.html "An unsigned integer type used for the number of file descriptors." I guess there's no reason to make it long though.

Reply:
`int` makes gcc complain about different signage. I think Kernel can't linclude LibC (?); at least several other structs don't use the LibC typedefs. Moved it from unsigned long to unsigned.

Reply:
The kernel can't include LibC, but we have `UnixTypes.h` for that

Reply:
Indeed, although let's be honest, it's pretty smelly to have so many things duplicated in `UnixTypes.h`.
Maybe one day we will have some kind of common type header shareable between the Kernel and LibC..

Reply:
With UnixTypes it can be an nfds_t in Process.cpp but not in Syscall.h as far as I understand. Is that what you want?